linfa-ensemble/src/random_forest/algorithm.rs

```diff
@@ -0,0 +1,122 @@
+use linfa_trees::DecisionTree;
+use crate::random_forest::hyperparameters::RandomForestParams;
+use ndarray::{Array1, Array2, ArrayBase, Data, Ix1, Ix2};
+use ndarray_rand::rand_distr::Uniform;
+use ndarray::Axis;
+use ndarray::Array;
+use ndarray_rand::RandomExt;
+use std::collections::HashMap;
+
+
+pub struct RandomForest {
+    pub hyperparameters: RandomForestParams,
+    pub trees: Vec<DecisionTree>,
+}
+
+impl RandomForest {
+    pub fn fit(
+        hyperparameters: RandomForestParams,
+        x: &ArrayBase<impl Data<Elem = f64>, Ix2>,
+        y: &ArrayBase<impl Data<Elem = u64>, Ix1>,
+    ) -> Self {
+
+    let n_estimators = hyperparameters.n_estimators;
+    let mut trees: Vec<DecisionTree> = Vec::with_capacity(n_estimators);
+    let single_tree_params = hyperparameters.tree_hyperparameters;
+
+    //TODO check bootstrap
+    let _bootstrap = hyperparameters.bootstrap;
+
+    for _ in 0..n_estimators {
+        // Bagging here
+        let rnd_idx = Array::random((1, x.nrows()), Uniform::new(0, x.nrows())).into_raw_vec();
+        let xsample = x.select(Axis(0), &rnd_idx);
+        let ysample = y.select(Axis(0), &rnd_idx);
+
+        let tree = DecisionTree::fit(single_tree_params, &xsample, &ysample);
+        trees.push(tree);
+    }
+
+    Self {
+        hyperparameters: hyperparameters,
+        trees: trees
+        }
+    }
+
+    pub fn predict(&self, x: &ArrayBase<impl Data<Elem = f64>, Ix2>) -> Array1<u64> {
+        let ntrees = self.hyperparameters.n_estimators;
+        assert!(ntrees > 0, "Run .fit() method first");
+
+        let mut predictions: Array2<u64> = Array2::zeros((ntrees, x.nrows()));
+
+        for i in 0..ntrees {
+            let single_pred = self.trees[i].predict(&x);
+            dbg!("single pred: ", &single_pred);
+
+            // TODO can we make this more idiomatic rust
+            for j in 0..single_pred.len() {
+                predictions[[i, j]] = single_pred[j];
```

Comment:
I don't know whether you can collect to `Array2`, but it's easier to write that with iterators.
```rust
self.trees.iter().map(|tree| tree.predict(&x)).collect::<Array2<_>>()
```
otherwise you can do something like [this](https://stackoverflow.com/questions/53698503/how-to-convert-a-collection-of-vecndarrayarray1-into-an-array2)